data/transactions/verify/verifiedTxnCache.go

```diff
@@ -0,0 +1,292 @@
+					if params, has := v.buckets[bucketIdx][id]; has {
+						entryGroup = params
+						baseBucket = bucketIdx
+						break
+					}
+				}
+			}
+
+			if entryGroup == nil {
+				break
+			}
+
+			if !entryGroup.Equal(groupCtx) {
+				break
+			}
+
+			if entryGroup.signedGroupTxns[txnIdx].Sig != txn.Sig || (!entryGroup.signedGroupTxns[txnIdx].Msig.Equal(txn.Msig)) || (!entryGroup.signedGroupTxns[txnIdx].Lsig.Equal(&txn.Lsig)) || (entryGroup.signedGroupTxns[txnIdx].AuthAddr != txn.AuthAddr) {
+				break
+			}
+			verifiedTxn++
+		}
+		if verifiedTxn != len(signedTxnGroup) || verifiedTxn == 0 {
+			unverifiedGroups = append(unverifiedGroups, signedTxnGroup)
+		}
+	}
+	return
+}
+
+// UpdatePinned replaces the pinned entries with the one provided in the pinnedTxns map. This is typically expected to be a subset of the
+// already-pinned transactions. If a transaction is not currently pinned, and it's can't be found in the cache, a errMissingPinnedEntry error would be generated.
+func (v *verifiedTransactionCache) UpdatePinned(pinnedTxns map[transactions.Txid]transactions.SignedTxn) (err error) {
+	v.bucketsLock.Lock()
+	defer v.bucketsLock.Unlock()
+	pinned := make(map[transactions.Txid]*GroupContext, len(pinnedTxns))
+	for txID := range pinnedTxns {
+		if groupEntry, has := v.pinned[txID]; has {
+			pinned[txID] = groupEntry
+			continue
+		}
+
+		// entry isn't in pinned; maybe we have it in one of the buckets ?
+		found := false
+		// we use the (base + W) % W trick here so we can go backward and wrap around the zero.
+		for offsetBucketIdx := v.base + len(v.buckets); offsetBucketIdx > v.base; offsetBucketIdx-- {
+			bucketIdx := offsetBucketIdx % len(v.buckets)
+			if groupEntry, has := v.buckets[bucketIdx][txID]; has {
+				pinned[txID] = groupEntry
+				found = true
+				break
+			}
+		}
+		if !found {
+			err = errMissingPinnedEntry
+		}
+
+	}
+	v.pinned = pinned
+	return err
+}
+
+// Pin sets a given transaction group as pinned items, after they have already been verified.
+func (v *verifiedTransactionCache) Pin(txgroup []transactions.SignedTxn) (err error) {
+	v.bucketsLock.Lock()
+	defer v.bucketsLock.Unlock()
+	transcationMissing := false
+	if len(v.pinned)+len(txgroup) > maxPinnedEntries {
+		// reaching this number likely means that we have an issue not removing entries from the pinned map.
+		// return an error ( which would get logged )
+		return errTooManyPinnedEntries
+	}
+	baseBucket := v.base
+	for _, txn := range txgroup {
+		txID := txn.ID()
+		if _, has := v.pinned[txID]; has {
+			// it's already pinned; keep going.
+			continue
+		}
+
+		// entry isn't in pinned; maybe we have it in one of the buckets ?
+		found := false
+		// we use the (base + W) % W trick here so we can go backward and wrap around the zero.
+		for offsetBucketIdx := baseBucket + len(v.buckets); offsetBucketIdx > baseBucket; offsetBucketIdx-- {
+			bucketIdx := offsetBucketIdx % len(v.buckets)
+			if ctx, has := v.buckets[bucketIdx][txID]; has {
```

Comment:
nit: we might stop earlier if we track how many buckets are in use. Maybe not a big deal, will only help on non-full cache.

Reply:
I think that after the first cycle, all the bucket will be in use ( although they might contain "old" entries ).
My intent here was to try and avoid deleting the old maps entries.